lib/block_view_atomic_txns.go

```diff
@@ -0,0 +1,476 @@
+	delete(msgDuplicate.ExtraData, PreviousAtomicTxnPreHash)
+
+	// Convert the transaction to bytes but do NOT encode the transaction signature.
+	preSignature := true
+	txBytes, err := msgDuplicate.ToBytes(preSignature)
+	if err != nil {
+		return nil, errors.Wrap(err, "MsgDeSoTxn.AtomicHash: cannot convert modified transaction to bytes")
+	}
+
+	// Return the SHA256 double hash of the resulting bytes.
+	return Sha256DoubleHash(txBytes), nil
+}
+
+//
+// Connect and Disconnect Atomic Txn Logic
+//
+
+func (bav *UtxoView) _connectAtomicTxnsWrapper(
+	txn *MsgDeSoTxn,
+	txHash *BlockHash,
+	blockHeight uint32,
+	blockTimestampNanoSecs int64,
+	verifySignatures bool,
+	ignoreUtxos bool,
+) (
+	_utxoOps []*UtxoOperation,
+	_totalInput uint64,
+	_totalOutput uint64,
+	_fees uint64,
+	_err error,
+) {
+	var utxoOpsForTxn []*UtxoOperation
+
+	// Don't allow the atomic transactions and the wrapper to take up more than half of the block.
+	txnBytes, err := txn.ToBytes(false)
+	if err != nil {
+		return nil, 0, 0, 0, errors.Wrapf(
+			err, "_connectTransaction: Problem serializing transaction: ")
+	}
+	txnSizeBytes := uint64(len(txnBytes))
+	if txnSizeBytes > bav.Params.MaxBlockSizeBytes/2 {
+		return nil, 0, 0, 0, RuleErrorTxnTooBig
+	}
+
+	// Validate the connecting block height.
+	if blockHeight < bav.Params.ForkHeights.ProofOfStake1StateSetupBlockHeight {
+		return nil, 0, 0, 0,
+			errors.Wrap(RuleErrorAtomicTxnBeforeBlockHeight, "_connectAtomicTxnsWrapper")
+	}
+
+	// Validate the transaction metadata type.
+	if txn.TxnMeta.GetTxnType() != TxnTypeAtomicTxnsWrapper {
+		return nil, 0, 0, 0,
+			fmt.Errorf("_connectAtomicTxnsWrapper: TxnMeta type: %v", txn.TxnMeta.GetTxnType().GetTxnString())
+	}
+
+	// Verify the wrapper of the transaction. This does not verify the txn.TxnMeta contents, just that
+	// the wrapper is well formatted.
+	if err := _verifyAtomicTxnsWrapper(txn); err != nil {
+		return nil, 0, 0, 0, errors.Wrap(err, "_connectAtomicTxnsWrapper")
+	}
+
+	// Extract the metadata from the transaction.
+	txMeta := txn.TxnMeta.(*AtomicTxnsWrapperMetadata)
+
+	// Verify the chain of transactions as being not tampered with. This verifies the txn.TxnMeta contents.
+	if err := _verifyAtomicTxnsChain(txMeta); err != nil {
+		return nil, 0, 0, 0, errors.Wrap(err, "_connectAtomicTxnsWrapper")
+	}
+
+	// Execute the internal transactions.
+	var innerUtxoOps [][]*UtxoOperation
+	var totalInput, totalOutput, totalFees uint64
+	for _, innerTxn := range txMeta.Txns {
+		// NOTE: By recursively calling _connectNonAtomicTransaction, each inner transaction is checked that
+		// it is capable of paying for its own fees as well as having a valid signature.
+		innerTxnUtxoOps, txnInput, txnOutput, txnFees, err := bav._connectNonAtomicTransaction(
+			innerTxn, txHash, blockHeight, blockTimestampNanoSecs, verifySignatures, ignoreUtxos)
+		if err != nil {
+			return nil, 0, 0, 0,
+				errors.Wrap(err, "_connectAtomicTxnsWrapper")
```

Comment:
Usually error should be descriptive and end in ": " when using Wrap. See other examples.